perses/rjmc/topology_proposal.py

```diff
@@ -401,46 +423,82 @@ def _get_mol_atom_map(current_oemol,
+                        map_score += all_to_all[M[atom], atom]
+                all_scores.append(map_score/len(M))
+            _logger.debug(f'Mapping scores: {all_scores}')
+            # TODO: return one map from any group of scores
+
+            # returning lowest score
+            best_map_index = np.argmin(all_scores)
+            _logger.debug(f'Returning map index: {best_map_index}')
+            return all_new_to_old_atom_maps[best_map_index]
 
-        #now all else is equal; we will choose the map with the highest overlap of atom indices
-        index_overlap_numbers = []
-        if matching_criterion == 'index':
-            for map in new_to_old_atom_maps:
-                hit_number = 0
-                for key, value in map.items():
-                    if key == value:
-                        hit_number += 1
-                index_overlap_numbers.append(hit_number)
-        elif matching_criterion == 'name':
-            for map in new_to_old_atom_maps:
-                hit_number = 0
-                map_tuples = list(map.items())
-                atom_map = {atom_new: atom_old for atom_new, atom_old in zip(list(proposed_oemol.GetAtoms()), list(current_oemol.GetAtoms())) if (atom_new.GetIdx(), atom_old.GetIdx()) in map_tuples}
-                for key, value in atom_map.items():
-                    if key.GetName() == value.GetName():
-                        hit_number += 1
-                index_overlap_numbers.append(hit_number)
         else:
-            raise Exception(f"the ranking criteria {ranking_criteria} is not supported.")
+            new_to_old_atom_maps = [map for count, map in zip(count_after_hydrogen_mapping, all_new_to_old_atom_maps) if count == max_num_atoms]
+
+        # now all else is equal; we will choose the map with the highest overlap of atom indices
+            index_overlap_numbers = []
+            if matching_criterion == 'index':
+                for map in new_to_old_atom_maps:
+                    hit_number = 0
+                    for key, value in map.items():
+                        if key == value:
+                            hit_number += 1
+                    index_overlap_numbers.append(hit_number)
+            elif matching_criterion == 'name':
+                for map in new_to_old_atom_maps:
+                    hit_number = 0
+                    map_tuples = list(map.items())
+                    atom_map = {atom_new: atom_old for atom_new, atom_old in zip(list(proposed_oemol.GetAtoms()), list(current_oemol.GetAtoms())) if (atom_new.GetIdx(), atom_old.GetIdx()) in map_tuples}
+                    for key, value in atom_map.items():
+                        if key.GetName() == value.GetName():
+                            hit_number += 1
+                    index_overlap_numbers.append(hit_number)
+            else:
+                raise Exception(f"the ranking criteria {matching_criterion} is not supported.")
```

Comment:
so the geometry filter is happening before before mapping of atomic indices filter...would this run faster if we filtered by geometry _after_ everything else?

Reply:
Yes - unless geometry='weak'. Doing mapping _then_ geometry doesn't work for the CDK2 example I explain above, because of the hydrogen mapping making the top matches wrong.